common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java

```diff
@@ -171,7 +171,9 @@ private void failRemainingBlocks(String[] failedBlockIds, Throwable e) {
 
     @Override
     public void onData(String streamId, ByteBuffer buf) throws IOException {
-      channel.write(buf);
+      while (buf.hasRemaining()) {
+        channel.write(buf);
```

Comment:
I actually thought this is OK for a FileChannel (just based on this comment in the original: https://github.com/apache/spark/pull/16989/files#r115409001)

but certainly this seems like a safe change.  Just wondering if you observed an issue, or this is just to be safe.

Reply:
I don't hit this issue and it may be just because this feature is disabled by default.

Just to be safe as [FileDispatcherImpl](http://www.grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/sun/nio/ch/FileChannelImpl.java#210) calls `write/pwrite` system call but doesn't check the written size. Seems the written size may be less than the buffer size: https://stackoverflow.com/questions/32683086/handling-incomplete-write-calls

Reply:
[FileSuite.writeBinaryData](https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/FileSuite.scala#L247) this also should be fixed?

Reply:
@ConeyLiu Good catch. Let me also fix it.